src/main/java/com/puppycrawl/tools/checkstyle/filters/XpathFilter.java

```diff
@@ -115,6 +115,56 @@ public XpathFilter(String files, String checks,
         }
     }
 
+    /**
+     * Creates a {@code XpathElement} instance.
+     * @param files regular expression for names of filtered files
+     * @param checks regular expression for filtered check classes
+     * @param message regular expression for messages.
+     * @param moduleId the module id
+     * @param query the xpath query
+     */
+    public XpathFilter(Pattern files, Pattern checks,
+                       Pattern message, String moduleId, String query) {
+        if (files == null) {
+            filePattern = null;
+            fileRegexp = null;
+        }
+        else {
+            filePattern = files.pattern();
```

Comment:
I am not quite sure what you mean. to drop the String field. So should I drop fields such as filePattern and re-write the code in hashCode and equals methods?

Reply:
I tried to do so. However, I found that when the value of filePattern comes to null. I cannot use the .pattern() method in hashCode and equals methods which will throw the null pointer exception. Unless I add more if statement in these both methods to prevent the null Pattern.

Reply:
The other way I think about is always preventing the Pattern value becomes null. When users didn't put any value in these Pattern fields to suppress, it would compile the default value which is ".*" in regexp, which means that I will re-write the if statements of setters in SuppressionXpathSingleFilter. e.g.
```java
public void setChecks(String checks) {
        if (checks == null) {
            this.checks = Pattern.compile(".*");
        }
        else {
            this.checks = Pattern.compile(checks);
        }
    }
```
Please give some advices, thx.

Reply:
>  I cannot use the .pattern() method to get the pattern of it in hashCode and equals methods which will throw the null pointer exception.

Why can't you use the Pattern class directly? `Objects.hash` and `Objects.equals` will protect you from null variables, and if you give it the pattern class directly, then you won't need to call the pattern method.

Reply:
Based on https://github.com/checkstyle/checkstyle/pull/6172#issuecomment-443866697 , this can't be done because Pattern doesn't implement its own hashCode or equals methods. So it see 2 different objects with the same pattern as different.
I will mark this as resolved.